src/main/java/htsjdk/samtools/util/AsyncWriterPool.java

```diff
@@ -0,0 +1,126 @@
+    }
+
+    @Override
+    public void close() throws IOException {
+        this.poolClosed.set(true);
+        for (PooledWriter<?> writer : this.writers) {
+            writer.close();
+        }
+        this.executor.shutdown();
+    }
+
+
+    // TODO: note that this is not threadsafe and multiple threads trying to call write or close should by synchronized
+    public class PooledWriter<A> implements Writer<A> {
+        private final Writer<A> writer;
+        private final BlockingQueue<A> queue;
+        private final AtomicBoolean writing;
+        private final AtomicBoolean isClosed;
+
+        private final int bufferSize;
+
+        // Holds the Future of the last task submitted to the AsyncWriterPools executor, or null if nothing has been
+        // submitted yet. Inside the Future is the number of items written and any exceptions that may have been thrown.
+        private Future<Void> currentTask;
+
+
+        public PooledWriter(final Writer<A> writer, final BlockingQueue<A> queue, final int bufferSize) {
+            assert bufferSize >= 1; // TODO: Add error message
+            this.writer = writer;
+            this.queue = queue;
+            this.bufferSize = bufferSize;
+            this.writing = new AtomicBoolean(false);
+            this.isClosed = new AtomicBoolean(false);
+            this.currentTask = null;
+
+            // Add to pools writers
+            AsyncWriterPool.this.writers.add(this);
+        }
+
+        /**
+         * Will throw any error thrown in task
+         */
+        private void checkAndRethrow() {
+            if (this.currentTask != null) {
+                try {
+                    // NB: ignoring output of number of records previously written
+                    this.currentTask.get();
+                } catch (CancellationException | InterruptedException | ExecutionException e) {
+                    this.isClosed.set(true);
+                    throw new RuntimeException("Exception while writing records asynchronously", e);
+                }
+            }
+        }
+
+        @Override
+        public void write(final A item) {
+            if (this.isClosed.get()) throw new RuntimeIOException("Attempt to add record to closed writer.");
+
+            // Put new item in queue
+            try {
+                this.queue.put(item);
+            } catch (InterruptedException e) {
+                throw new RuntimeException("Exception while placing item in queue", e);
+            }
+
+            // If queue size is large enough, and not currently writing items, send to executor
+            if (queue.size() >= this.bufferSize && !this.writing.getAndSet(true)) {
+
+                checkAndRethrow();
+                this.drain(this.queue.size());
+            }
+        }
+
+        private void drain(int toDrain) {
+            // check the result of the previous task
+            this.currentTask = AsyncWriterPool.this.executor.submit(() -> {
```

Comment:
So far, running with an unbounded queue, this has not been the case, and since the caller can supply the queue type I'm not too worried about it.

It is possible, but given that I'm now draining queues till empty, even if there is a pathological input dataset that is sending to just one writer, we become a pass-through writer on a different thread, writing as fast as it's adding data with just a bit of extra overhead for copying to the intermediate queue. So I think it should be fine.

Reply:
Perhaps a note to here is appropriate to say this could theoretically happen but hasn't been observed practically.